core/services/balance_monitor.go

```diff
@@ -0,0 +1,129 @@
+type BalanceMonitor interface {
+	store.HeadTrackable
+}
+
+type balanceMonitor struct {
+	store             *store.Store
+	gethClientWrapper store.GethClientWrapper
+	ethBalances       map[gethCommon.Address]*assets.Eth
+	ethBalancesMtx    *sync.Mutex
+}
+
+// NewBalanceMonitor returns a new balanceMonitor
+func NewBalanceMonitor(store *store.Store, gethClientWrapper store.GethClientWrapper) BalanceMonitor {
+	return &balanceMonitor{
+		store:             store,
+		gethClientWrapper: gethClientWrapper,
+		ethBalances:       make(map[gethCommon.Address]*assets.Eth),
+		ethBalancesMtx:    new(sync.Mutex),
+	}
+}
+
+// Connect complies with HeadTrackable
+func (bm *balanceMonitor) Connect(_ *models.Head) error {
+	// Connect head can be out of date, so always query the latest balance
+	bm.checkBalance(nil)
+	return nil
+}
+
+// Disconnect complies with HeadTrackable
+func (bm *balanceMonitor) Disconnect() {}
+
+const ethFetchTimeout = 2 * time.Second
+
+// OnNewLongestChain checks the balance for each key
+func (bm *balanceMonitor) OnNewLongestChain(head models.Head) {
+	bm.checkBalance(&head)
+}
+
+func (bm *balanceMonitor) checkBalance(head *models.Head) {
+	keys, err := bm.store.Keys()
+	if err != nil {
+		logger.Error("BalanceMonitor: error getting keys", err)
+	}
+
+	var wg sync.WaitGroup
+
+	for _, key := range keys {
+		wg.Add(1)
+
+		go func(k models.Key) {
+			ctx, cancel := context.WithTimeout(context.Background(), ethFetchTimeout)
+			defer cancel()
+
+			bm.gethClientWrapper.GethClient(func(c eth.GethClient) error {
+				var err error
+				var bal *big.Int
+				var headNum *big.Int
+				if head == nil {
+					headNum = nil
+					bal, err = c.BalanceAt(ctx, k.Address.Address(), nil)
+				} else {
+					headNum = big.NewInt(head.Number)
+					bal, err = c.BalanceAt(ctx, k.Address.Address(), headNum)
+				}
+				if err != nil {
+					err := errors.Wrap(err, "error getting balance")
+					logger.Errorw(fmt.Sprintf("BalanceMonitor: error getting balance for key %s: %s", k.Address.Hex(), err.Error()), "err", err, "address", k.Address, "headNum", headNum)
+				}
+				ethBal := assets.NewEth(bal.Int64())
+
+				bm.updateBalance(ethBal, k.Address.Address(), headNum)
+				return nil
+			})
+
+			wg.Done()
+		}(key)
+	}
+
+	wg.Wait()
+}
+
+func (bm *balanceMonitor) updateBalance(ethBal *assets.Eth, address gethCommon.Address, headNum *big.Int) {
+	if ethBal == nil {
+		logger.Error("BalanceMonitor: invariant violation, ethBal should never be nil")
+		return
+	}
+	store.PromUpdateEthBalance(ethBal, address)
+
+	bm.ethBalancesMtx.Lock()
+	oldBal := bm.ethBalances[address]
+	bm.ethBalances[address] = ethBal
+	bm.ethBalancesMtx.Unlock()
+
+	loggerFields := []interface{}{
+		"address", address.Hex(),
+		"headNum", headNum,
+		"ethBalance", ethBal.String(),
+		"weiBalance", ethBal.ToInt(),
+	}
+
+	if oldBal == nil {
+		logger.Infow(fmt.Sprintf("ETH balance for %s: %s", address.Hex(), ethBal.String()), loggerFields...)
```

Comment:
Are node operators relying on these logs lines to trigger alerts? 
Do we need to make sure the error message has the same format as in the previous tx manager?

Reply:
Good question, I checked with Jonny and as long as the ethBalance key is present it ought to be fine.